clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp

```diff
@@ -0,0 +1,163 @@
+//===--- AutoCheck.cpp - clang-tidy ---------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AutoCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+void AutoCheck::registerMatchers(MatchFinder *Finder) {
+  // We register 3 matchers separately:
+  //  - Variable declarations (which includes parameters)
+  //  - Function declarations (to check return types)
+  //  - Lambdas (which are handled separately given they have slightly different
+  //  rules).
+
+  // This matcher should not bind to implicit variables added by clang, that is
+  // why it uses traverse with TK_IgnoreUnlessSpelledInSource.
+  const auto hasAutoType =
+      anyOf(hasType(autoType()), hasType(references(autoType())),
+            hasType(pointsTo(autoType())),
+            allOf(hasType(templateTypeParmType()), hasType(asString("auto"))));
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource, varDecl(hasAutoType))
+          .bind("variable-decl"),
+      this);
+  Finder->addMatcher(
+      functionDecl(returns(autoType()), unless(hasTrailingReturn()))
+          .bind("function-return"),
+      this);
+  Finder->addMatcher(cxxRecordDecl(isLambda()).bind("lambda"), this);
+}
+
+void AutoCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *MatchedDecl =
+          Result.Nodes.getNodeAs<VarDecl>("variable-decl")) {
+    std::string actualType{""};
+
+    // Check if it is a parameter part of a method, part of a closure record
+    // representing a lambda
+    if (const ParmVarDecl *paramVarDecl =
+            llvm::dyn_cast<ParmVarDecl>(MatchedDecl)) {
+      if (const CXXMethodDecl *method = llvm::dyn_cast<CXXMethodDecl>(
+              paramVarDecl->getParentFunctionOrMethod())) {
+        if (method->getParent()->isLambda()) {
+          // auto lambda parameters are allowed
+          return;
+        }
+      }
+    }
+
+    if (auto *autoType = llvm::dyn_cast<AutoType>(MatchedDecl->getType())) {
+      // Check if the auto variable contains a lambda
+      if (auto *recordType =
+              llvm::dyn_cast<RecordType>(autoType->getDeducedType())) {
+        if (CXXRecordDecl *recordDecl =
+                dyn_cast<CXXRecordDecl>(recordType->getDecl());
+            recordDecl && recordDecl->isLambda()) {
+          // It is an auto variable containing a lambda
+          return;
+        }
+      }
+
+      // If it is constrained, we allow the type (just like we would for a
+      // template or concept)
```

Comment:
Good idea. We need update our CS too

Reply:
Sounds good 👍🏻 I'm collecting all changes to the CS in another task, will add it there.